service/controller/chart/resource/release/update.go

```diff
@@ -229,5 +245,106 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release does not have to be updated", desiredReleaseState.Name))
 	}
 
+	err = r.deleteRollbackAnnotation(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
 	return nil, nil
 }
+
+func (r *Resource) rollback(ctx context.Context, obj interface{}, currentStatus string) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	counts, ok := cr.GetAnnotations()[annotation.RollbackCount]
+
+	var rollbackCount int
+
+	if !ok {
+		rollbackCount = 0
+	} else {
+		rollbackCount, err = strconv.Atoi(counts)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	if rollbackCount > r.maxRollback {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release is in status %#q and cannot be updated despite of %d rollbacks", key.ReleaseName(cr), currentStatus, r.maxRollback))
```

Comment:
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release is in status %#q and has reached %d rollbacks", key.ReleaseName(cr), currentStatus, r.maxRollback))
```

I tried to make this a bit shorter. WDYT?

Reply:
much better, thanks!